src/fs.rs

```diff
@@ -0,0 +1,184 @@
+//! # ByteCache File System Operations
+//!
+//! This module handles file system interactions for the ByteCache system,
+//! providing interfaces for reading from and writing to cache storage on disk.
+//!
+//! It implements two primary access patterns:
+//!
+//! * **Read-only access**: Used for sideloaded content that should not be modified
+//! * **Read-write access**: Used for the persistent disk cache
+//!
+//! The module uses a marker type pattern to distinguish between these access modes
+//! at compile time, ensuring that operations are only performed when appropriate.
+
+use std::path::PathBuf;
+
+/// Marker type for read-only filesystem operations.
+///
+/// This is used for cache layers that should only read pre-existing data,
+/// such as the sideload cache.
+pub struct Read;
```

Comment:
```suggestion
pub struct Read(());
```

This prevents constructing this type, which should never be constructed.